### PR TITLE
Implement nested types: Add StructDtype and StructArray

### LINKED ISSUE #40652: ENH: Pandas StructDType
I already searched for a while for discussions of nested structures in Pandas, but I couldn't find anything corresponding.

#### Is your feature request related to a problem?

Currently, there is no way to work with arbitrary nested data types in Pandas.
In Spark and PyArrow, one can have StructTypes. In NumPy, we have Compound types.
However, when we try something like this:
```python
df = pd.DataFrame({'pos':[1,2,3], "val": ["a","b", "c"]})
struct = df.to_records(index=False).view(type=np.ndarray, dtype=list(df.dtypes.items()))
pd.Series(struct)
```
Then we only get an error:
```python
ValueError: Cannot construct a Series from an ndarray with compound dtype.  Use DataFrame instead.
```

This would be useful for a number of use cases:
- Direct mapping between PySpark and Pandas nested columns
- Easy creation of custom data types; type checking of nested columns
- Efficient serialization with Arrow

#### Describe the solution you'd like

My wish would be to have a generic Pandas StructDType that:
- can be composed of any other Pandas DType
- allows conversion to PyArrow StructType back and forth
- can be written to Parquet

A perfect example is the `IntervalDType`/`IntervalArray` that is already implemented in Pandas:
https://github.com/pandas-dev/pandas/blob/2198f51911dd5a5a0cc1c12c3f083b86c96bd33b/pandas/core/arrays/interval.py#L146
In my opinion, its implementation is a special case of a Struct dtype.
It also supports conversion to and from PyArrow (see https://github.com/pandas-dev/pandas/commit/2198f51911dd5a5a0cc1c12c3f083b86c96bd33b).
Therefore, by generalizing the IntervalDType to use any number of subtypes, we would have the StructDType implementation ready.

#### API breaking implications

`to_csv()`, etc. could have difficulties with storing nested data.
That's maybe a followup problem to solve.

#### Describe alternatives you've considered

One can try to construct the Series as a list of tuples.
However, this has two drawbacks:
- No type checking
- `to_parquet()` fails

### PR DESCRIPTION
**Pull request by Hoeze**
This is my first attempt to implement a StructDtype that is based on NamedTuples.
My end-goal is to reach full interoperability with PyArrow/PySpark and deeply nested types.
In particular, this StructDtype and StructArray should make it much simpler to create custom data types.

Some implementation details:
- A `StructArray` consists of an `OrderedDict` ("field_name" -> "pandas_type") and a mask if it is nullable.
- The scalar type of the `StructDtype` is accessible via `dtype.type` and can be overridden with custom data types.
- In theory, all functions should be recursive s.t. a `StructDtype` can be arbitrarily nested. (I need to test this later.)

There are a number of open issues to resolve, but I'd be very happy about some profound review or tips for improving.
- I kept the dtype and array classes in a single python file to make reviews simpler for now. These will need to be split up into the correct files later.
- The lines https://github.com/Hoeze/pandas/blob/d651395143cc4cadbb7d18fae2f98ef2354710ab/pandas/core/arrays/struct.py#L42-L88 are very ugly hacks for some methods where I could not find the corresponding Python functions for:
  - How to allocate empty Pandas arrays?
  - Why does `pd.api.types.pandas_dtype(pd.api.types.infer_dtype(v))` not work for e.g. integers?
  - Why is there no standard way to concatenate Pandas arrays?
- How bad is it to have `NamedTuple` as scalar type? This makes it impossible to have e.g. "0" as field name.
- How would I overwrite `pa.array().to_pandas()` for any `pa.StructType`?
- Where should nullability be handled? In Pyarrow, a dtype itself is not nullable but an array or StructField can be. In pandas, it seems like nullability needs to be implemented in every array class again.

---

- [x] closes #40652
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/core/arrays/struct.py [hunk 1]
```diff
@@ -0,0 +1,1117 @@
...
+
+import numpy as np
+import pandas as pd
+from pandas.api.extensions import (
+    register_extension_dtype,
+    ExtensionDtype,
+    ExtensionArray,
+)
+
+# from pandas.core.construction import extract_array
```
Comment (jbrockmendel) [inline pandas/core/arrays/struct.py:20]: pls remove commented-out code wherever possible.  where not possible, pls comment as to why

File: pandas/core/arrays/struct.py [hunk 2]
```diff
@@ -0,0 +1,1117 @@
...
+
+try:
+    import pyarrow as pa
+except ImportError:
+    pa = None
+
+import logging
+
+log = logging.getLogger(__name__)
```
Comment (jbrockmendel) [inline pandas/core/arrays/struct.py:29]: is this used?

File: pandas/core/arrays/struct.py [hunk 3]
```diff
@@ -0,0 +1,1117 @@
...
+        - ref: array of reference sequences
+        - alt: array of alternative sequences
+
+    In addition, can be converted from/to VCF-formatted (chrom, pos, ref, alt) representations.
+    """
+    _fields: OrderedDict[str, ExtensionArray]
+    "fields that this StructArray holds"
+    # _field_names: List[str]
+    # "ordered list of field names that this StructArray holds"
+    _mask: Union[NoneType, np.ndarray]
```
Comment (jbrockmendel) [inline pandas/core/arrays/struct.py:310]: `_mask: np.ndarray | None`, then can remove NoneType

**Comment by jbrockmendel** [comment]
> How to allocate empty Pandas arrays?

im assuming you mean ExtensionArray, since PandasArray is a specific subclass.  Use `dtype.empty(shape)`

> Why does pd.api.types.pandas_dtype(pd.api.types.infer_dtype(v)) not work for e.g. integers?

It is a bit annoying that infer_dtype doesn't return a dtype or dtype-like-string.  You might be better off with `infer_dtype_from(v)`

> Why is there no standard way to concatenate Pandas arrays?

core.dtypes.concat.concat_compat

> In pandas, it seems like nullability needs to be implemented in every array class again.

I'm working on making MaskedArray a wrapper that can be used around arbitrary other arrays, but don't hold your breath.

**Comment by mroeschke** [comment]
It might make sense just to have a ArrowStructArray directly and use the pyarrow.StructType as the backing dtype?

File: pandas/core/arrays/struct.py (hunk 2 above)
Comment (Hoeze) [reply pandas/core/arrays/struct.py:29]: Actually not, I'll remove it

**Comment by jreback** [comment]
> It might make sense just to have a ArrowStructArray directly and use the pyarrow.StructType as the backing dtype?

yeah I agree, we are moving towards backing with arrow types rather than having new types 'roll their own'. This is the scalable longer term soln. There is already a ArrowBackedExtensionArray subclass that you can start from, and @mroeschke is doing some work on this to extend to numeric operations.

I would be +1 on an implementation (even if limitted) in this way.

**Comment by Hoeze** [comment]
Another thing I conclude from https://github.com/pola-rs/polars/pull/3007:
We should not use NamedTuple as baseclass but rather subclass tuple with a custom `__getitem__()`:
- `tuple[idx]` where `idx` is an integer -> positional indexing of the field
- `tuple[col]` where `col` is a string -> get field by name
This way, we can keep support for arbitrary column names.

I will continue working on this PR as far as my free time allows.

**Pull request closed without merging**